### PR TITLE
[Merged by Bors] - feat: tensor algebra of free module over integral domain is a domain

### PR DESCRIPTION
**Pull request by bustercopley**
Provide instances
  * `Nontrivial (TensorAlgebra R M)` when `M` is a module over
    a nontrivial semiring `R`
  * `NoZeroDivisors (FreeAlgebra R X)` when `R` is a commutative
    semiring with no zero-divisors and `X` any type
  * `IsDomain (FreeAlgebra R X)` when `R` is an integral domain
    and `X` is any type
  * `TwoUniqueProds (FreeMonoid X)` where `X` is any type
    (this provides `NoZeroDivisors (MonoidAlgebra R (FreeMonoid X))`
    when `R` is a semiring and `X` any type,
    via `TwoUniqueProds.toUniqueProds`
    and `MonoidAlgebra.instNoZeroDivisorsOfUniqueProds`)
  * `NoZeroDivisors (TensorAlgebra R M)` when `M` is a free module
    over a commutative semiring `R` with no zero-divisors
  * `IsDomain (TensorAlgebra R M)` when `M` is a free module over
    an integral domain `R`

In Algebra.Group.UniqueProds:
* Rename `UniqueProds.mulHom_image_of_injective` to
  `UniqueProds.of_injective_mulHom`.
* New lemmas `UniqueMul.of_mulHom_image`, `UniqueProds.of_mulHom`,
  `TwoUniqueProds.of_mulHom` show the relevant property holds in the
  domain of a multiplicative homomorphism if it holds in the codomain,
  under a certain hypothesis on the homomorphism.

### CODE REVIEW

File: Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean:77]: Is there a missing intermediate result about `FreeAlgebra` here?

File: Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean:63]: @adomani, I seem to recall you had a more general version of this for `MonoidAlgebra`?

File: Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean
Comment (bustercopley) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean:77]: Maybe! Here I used the maximality of the lengths of `xmax ∈ x.support` and `ymax ∈ y.support` to show that `xmax * ymax ∈ (x * y).support`. I'm not sure how the proof would go for `MonoidAlgebra R G` without some such length/degree/grade/valuation on `G`.

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean:77]: My claim here is that you are effectively using the fact that the `FreeAlgebra` has no zero divisors.

Comment (bustercopley) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean:77]: (I assume you weren't referring only to the part you highlighted? That is just `(e x) * (e y) = 0 ↔ e (x * y) = 0` for an `e: AlgEquiv`, by an easy combination of existing lemmas, done twice.)

Comment (bustercopley) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean:77]: > My claim here is that you are effectively using the fact that the `FreeAlgebra` has no zero divisors.

Sorry, my last message wasn't in reply to the above. Messages don't seem to be popping up in a timely manner.

But I think I coincidentally addressed your point? Those lines don't require `NoZeroDivisors`, if I'm not mistaken.

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean]: See the commit I just pushed to see what I meant :)

Comment (bustercopley) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean]: I updated the bulleted list in the commit comment.

File: Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean]: I'm thinking of https://github.com/leanprover-community/mathlib/pull/16299/files

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean]: I think the missing piece (which I think is true) is `UniqueProds (FreeMonoid X)`, from which everything else follows.

Comment (bustercopley) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean]: Wow! Yes, that's true (same proof, take maximal-`length` words). I'll write it up.

Comment (bustercopley) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean]: Done. Very cool, thank you!

Comment (bustercopley) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean]: Now `_root_.instNoZeroDivisorsMonoidAlgebraFreeMonoid` is gone, maybe the instances have better homes? Then there is no need for the new file `TensorAlgebra.Domain`.

`FreeMonoid.instUniqueProds` : FreeMonoid.Basic
`FreeAlgebra.instNoZeroDivisors` : FreeAlgebra.Basic
`TensorAlgebra.instNoZeroDivisors`, `TensorAlgebra.instDomain` : TensorAlgebra.Basis _[sic]_

File: Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean:37]: Is the stronger `TwoUniqueProds` true?

File: Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean]: Those look reasonable to me.

File: Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean
Comment (adomani) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean:37]: I just got back to my computer and it looks like I missed on a lot of fun here!

Anyway, I think that `TwoUniqueProds` should also hold: wherever the proof uses `max`, you could have used `min` just as well and this should be enough for `TwoUniqueProds` to go through.

File: Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean
Comment (bustercopley) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean]: (done)

File: Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean
Comment (bustercopley) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean]: Choose two words of _minimal_ length ... let me check that.

Comment (bustercopley) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean]: Yes, that'll work. Search-and-replace 'max' for 'min' and fix up a few things for an alternate proof of `UniqueProds`. That leaves a little housekeeping for edge cases. I'll have to leave that for another day, as it's getting late.

Comment (bustercopley) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean]: > I'll have to leave that for another day
... or stay up late and do it badly ...
Needs golf.

Comment (bustercopley) [reply Mathlib/LinearAlgebra/TensorAlgebra/Domain.lean]: > I just got back to my computer and it looks like I missed on a lot of fun here!
Yeah!
> Anyway, I think that `TwoUniqueProds` should also hold: wherever the proof uses `max`, you could have used `min` just as well and this should be enough for `TwoUniqueProds` to go through.
Believe it or not, I didn't see this message until just now.

File: Mathlib/LinearAlgebra/TensorAlgebra/Basis.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/TensorAlgebra/Basis.lean:76]: Can you add this for `FreeAlgebra` too?

File: Mathlib/Algebra/Group/UniqueProds.lean
Comment (eric-wieser) [inline Mathlib/Algebra/Group/UniqueProds.lean:629]:
```suggestion
  uniqueMul_of_one_lt_card {A} {B} h := by
```

**Review by eric-wieser — APPROVED** [review]
LGTM! I'll maybe give @adomani time to come give this a final look before merging

(and of course, you should look at it when it's not super late in case you had any more ideas)

**Comment by adomani** [comment]
This is all very nice, thanks!

I wonder whether this result should be more "generic" in terms of `(Two)UniqueProds` for Lists, Multisets, Finsets, as [TwoUniqueProds.instTwoUniqueProdsProdInstMul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/UniqueProds.html#TwoUniqueProds.instTwoUniqueProdsProdInstMul) works for products.

In fact, using a Lex order on Lists and an arbitrary linear order on the underlying type `X` should give an argument very similar to the one for the MVPolynomials result.

**Comment by eric-wieser** [comment]
> I wonder whether this result should be more "generic" in terms of `(Two)UniqueProds` for Lists, Multisets, Finsets, as [TwoUniqueProds.instTwoUniqueProdsProdInstMul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/UniqueProds.html#TwoUniqueProds.instTwoUniqueProdsProdInstMul) works for products.

This should indeed work for `TwoUniqueSums` and `Multiset`, but none of the other types have an `Add` or `Mul` with the right semantics, so I don't think this can work there.

**Comment by eric-wieser** [comment]
I tried that, but it needs a linear order.

**Comment by adomani** [comment]
I'll try this when I have some more time.  What I had in mind was to use a lex ordering on Lists and any linear order on the base type `X`, i.e. one that you obtain via a LinearExtension.

In the multiset case, I think that you are also extracting both the min and the max for a lex order on any linear extension.  Anyway, I may have some time to try this out tonight or tomorrow, but this is just a way of golfing further the PR: I am happy with the results, even if I think that the proofs can be streamlined.

**Comment by adomani** [comment]
This is what I had in mind, but it requires an extra import:
```lean
import Mathlib.Order.Extension.Linear

@[to_additive "Any `FreeAddMonoid` has the `TwoUniqueSums` property."]
instance FreeMonoid.instTwoUniqueProds {κ : Type*} : TwoUniqueProds (FreeMonoid κ) :=
  -- put an arbitrary linear order on `κ`
  let _ : LE κ := { le := (· = ·) }
  have : PartialOrder κ := {
    le_refl := fun _ => rfl
    le_trans := fun _ _ _ ↦ Eq.trans_le
    le_antisymm := fun _ _ a _ ↦ a }
  let _ : LinearOrder κ := inferInstanceAs (LinearOrder (LinearExtension κ))
  -- use the lexicographic order on `List κ` induced by the order on `κ`
  let LO : LinearOrder (FreeMonoid κ) := List.instLinearOrderList
  -- show that a `CovariantClass` holds for `List`
  have : CovariantClass (FreeMonoid κ) (FreeMonoid κ) HMul.hMul LO.lt :=
    ⟨fun a b c bc => List.Lex.append_left _ bc .. ⟩
  TwoUniqueProds.of_covariant_right
```

File: Mathlib/LinearAlgebra/TensorAlgebra/Basis.lean
Comment (bustercopley) [reply Mathlib/LinearAlgebra/TensorAlgebra/Basis.lean:76]: 99eecd81c1

**Comment by eric-wieser** [comment]
I'd argue that that approach is backwards; we shouldn't have to put a weird order on `FreeMonoid`, there should be a  lemma stated in terms of a suitable map `f : FreeMonoid _ -> Y`, where the order goes on `Y`.

**Comment by adomani** [comment]
> I'd argue that that approach is backwards; we shouldn't have to put a weird order on `FreeMonoid`, there should be a lemma stated in terms of a suitable map `f : FreeMonoid _ -> Y`, where the order goes on `Y`.

I always find that this issue with random orders appearing is probably due to my initial input to `UniqueProds`.  This comes from the fact that a standard way of *proving* that `UniqueProds` holds for `X`, is to put an order on `X` and extract mins or maxs from `Finset`s of `X`.

However, the statement of `UniqueProds` makes no reference to a background order.  All of this is to say that, when you are putting an arbitrary order on something, just to prove `UniqueProds`, it is probably an indication that a previous lemma that I put in the library assumed an order, but it could have been generalised to something without an order!

**Comment by adomani** [comment]
Anyway, I think that this discussion highlights the need for an improvement of the basics of the `UniqueProds` file.  This improvement, however, is out of scope for this PR.

I am happy with the current PR!

**Review by adomani — APPROVED** [review]

**Comment by bustercopley** [comment]
There's a trivial refactor I can do:
```
@[to_additive]
theorem of_graded_mul {X G : Type*} [Mul X] [Add G] [IsRightCancelAdd G] [LinearOrder G]
    [CovariantClass G G (· + ·) (· < ·)] [CovariantClass G G (Function.swap (· + ·)) (· < ·)]
    {f : X → G} (hf : ∀ x y : X, f (x * y) = f x + f y)
    (hinj : ∀ a b c d : X, a * b = c * d → f a = f c → a = c ∧ b = d) :
    TwoUniqueProds X where
  uniqueMul_of_one_lt_card {A} {B} h := by
    sorry
```
and
```
/-- Any `FreeMonoid` has the `TwoUniqueProds` property. -/
@[to_additive "Any `FreeAddMonoid` has the `TwoUniqueSums` property."]
instance FreeMonoid.instTwoUniqueProds {κ : Type*} : TwoUniqueProds (FreeMonoid κ) :=
  TwoUniqueProds.of_graded_mul (List.length_append) (fun _ _ _ _ => by exact List.append_inj)
```
It doesn't seem terribly useful. Maybe there's a way to do it that fits in with the API around `GradedMonoid`.

**Comment by eric-wieser** [comment]
> There's a trivial refactor I can do:

That looks like a nice refactor to me.

**Comment by bustercopley** [comment]
> > There's a trivial refactor I can do:
> 
> That looks like a nice refactor to me.

79faa1b10c

I'm not sure "graded" is precisely the word I want here. It's suggestive, at least.

**Comment by bustercopley** [comment]
Reviewers, thank you very much for your work on this. It's been an education.

Do you have any further suggestions?

* Do you think "grade" is ok? Can you see a better phrasing?
* More docstrings?
* All the `rw`'s should be `rewrite`'s, since the implicit `with_reducible rfl` is just a waste of time
  * note: the saving, in heartbeats, would be 67 / 3566 (1.9%)
  * note: I invite you to pass over this point without comment, inarguable though it may be; I'm a C++ programmer at heart, and we love our obsessive parsimony 
  * note: every occurrence of "`rewrite`" in current Mathlib is by me

**Comment by bustercopley** [comment]
No more improvements have come to mind. @eric-wieser, any suggestions? Do you think this should be merged, and do you think it's ready? Anything else I should do (like ask another person to review)?

**Comment by alreadydone** [comment]
@adomani Do you remember why you called this lemma `mulHom_image_of_injective` in [#6723](https://github.com/leanprover-community/mathlib4/pull/6723/commits/bebb7d8cba23140f843a053da8cd419facb9e896#diff-ebdf20b185bac5adc5a22ffdf53e7dfa74bb23fb34dfa47a761bb9dfa66c9fc1R233-R234)? I can't see `image` in the statement and I think `of_injective_mulHom` might be a better name.

**Comment by alreadydone** [comment]
I've completed [a refactor](https://github.com/leanprover-community/mathlib4/compare/richard-copley/IsDomainTensorAlgebra...jyxu/IsDomainTensorAlgebra?expand=1) cuts off 18 lines and proves the more general
```
@[to_additive] theorem of_mulHom (f : H →ₙ* G)
    (hf : ∀ ⦃a b c d : H⦄, a * b = c * d → f a = f c ∧ f b = f d → a = c ∧ b = d)
    (uG : TwoUniqueProds G) : TwoUniqueProds H where
```

**Comment by bustercopley** [comment]
Merged. `of_mulHom` is great!

**Comment by adomani** [comment]
> @adomani Do you remember why you called this lemma `mulHom_image_of_injective` in [#6723](https://github.com/leanprover-community/mathlib4/pull/6723/commits/bebb7d8cba23140f843a053da8cd419facb9e896#diff-ebdf20b185bac5adc5a22ffdf53e7dfa74bb23fb34dfa47a761bb9dfa66c9fc1R233-R234)? I can't see `image` in the statement (but the proof uses `mulHom_image_iff`) and I think `of_injective_mulHom` might be a better name.
> 
> ```
> theorem mulHom_image_of_injective (f : H →ₙ* G) (hf : Function.Injective f) (uG : UniqueProds G) :
>     UniqueProds H 
> ```

I think that your name is indeed better.  I suspect that the name that I gave referred to an older version of the lemma that *might* have contained an image...

**Review by alreadydone — APPROVED** [review]
Thanks! I wouldn't have thought of the condition `hf` without you extracting `of_graded_mul`; my `hf` is simply a weakened form of your `hinj` that assumes `f b = f d` in addition. What I observed is that in
```
theorem of_graded_mul {X G : Type*} [Mul X] [Add G] [IsRightCancelAdd G] [LinearOrder G]
    [CovariantClass G G (· + ·) (· < ·)] [CovariantClass G G (Function.swap (· + ·)) (· < ·)]
    {f : X → G} (hf : ∀ x y : X, f (x * y) = f x + f y)
    (hinj : ∀ a b c d : X, a * b = c * d → f a = f c → a = c ∧ b = d) :
    TwoUniqueProds X where
```
(let's pretend everything is multiplicative) `hf` is saying that `f` is a MulHom, and the conditions on G guarantees that G has TwoUniqueProds by [TwoUniqueProds.of_covariant_right](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/UniqueProds.html#TwoUniqueProds.of_covariant_right). Once I realized these it's natural to come up with `of_mulHom`.
This PR looks good to me now, but I just noticed no one ever added any labels to it? I added two and I'll wait a day for comments before sending it to maintainers.

File: Mathlib/Algebra/Group/UniqueProds.lean
Comment (alreadydone) [inline Mathlib/Algebra/Group/UniqueProds.lean:341]: Since we renamed `mulHom_image_of_injective`, I've now also renamed `mulHom_image_iff` for consistency.

File: Mathlib/Algebra/Group/UniqueProds.lean
Comment (alreadydone) [inline Mathlib/Algebra/Group/UniqueProds.lean:635]: And added deprecated aliases. (I doubt anyone is using them, but still ...)

**Comment by alreadydone** [comment]
maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
